velox/dwio/common/SelectiveBufferedInput.cpp

```diff
@@ -0,0 +1,297 @@
+        request.processed = true;
+        storageLoad.push_back(&request);
+      }
+    }
+    makeLoads(std::move(storageLoad), isPrefetchPct(readPct));
+  }
+}
+
+void SelectiveBufferedInput::makeLoads(
+    std::vector<LoadRequest*> requests,
+    bool prefetch) {
+  if (requests.empty() || (requests.size() < 2 && !prefetch)) {
+    return;
+  }
+  int32_t maxDistance = options_.maxCoalesceDistance();
+  auto loadQuantum = options_.loadQuantum();
+  // If reading densely accessed, coalesce into large for best throughput, if
+  // for sparse, coalesce to quantum to reduce overread. Not all sparse access
+  // is correlated.
+  auto maxCoalesceBytes = prefetch ? options_.maxCoalesceBytes() : loadQuantum;
+  std::sort(
+      requests.begin(),
+      requests.end(),
+      [&](const LoadRequest* left, const LoadRequest* right) {
+        return left->region.offset < right->region.offset;
+      });
+  // Combine adjacent short reads.
+
+  int32_t numNewLoads = 0;
+  int64_t coalescedBytes = 0;
+  coalesceIo<LoadRequest*, LoadRequest*>(
+      requests,
+      maxDistance,
+      // Break batches up. Better load more short ones i parallel.
+      1000, // limit coalesce by size, not count.
+      [&](int32_t index) { return requests[index]->region.offset; },
+      [&](int32_t index) -> int32_t {
+        auto size = requests[index]->region.length;
+        if (size > loadQuantum) {
+          coalescedBytes += loadQuantum;
+          return loadQuantum;
+        }
+        coalescedBytes += size;
+        return size;
+      },
+      [&](int32_t index) {
+        if (coalescedBytes > maxCoalesceBytes) {
+          coalescedBytes = 0;
+          return kNoCoalesce;
+        }
+        return 1;
+      },
+      [&](LoadRequest* request, std::vector<LoadRequest*>& ranges) {
+        ranges.push_back(request);
+      },
+      [&](int32_t /*gap*/, std::vector<LoadRequest*> /*ranges*/) { /*no op*/ },
+      [&](const std::vector<LoadRequest*>& /*requests*/,
+          int32_t /*begin*/,
+          int32_t /*end*/,
+          uint64_t /*offset*/,
+          const std::vector<LoadRequest*>& ranges) {
+        ++numNewLoads;
+        readRegion(ranges, prefetch);
+      });
+  if (prefetch && executor_) {
+    for (auto i = 0; i < allCoalescedLoads_.size(); ++i) {
+      auto& load = allCoalescedLoads_[i];
+      if (load->state() == CoalescedLoad::State::kPlanned) {
+        prefetchSize_ += load->size();
+        executor_->add([pendingLoad = load]() {
+          process::TraceContext trace("Read Ahead");
+          pendingLoad->loadOrFuture(nullptr);
+        });
+      }
+    }
+  }
+}
+
+void SelectiveBufferedInput::readRegion(
+    std::vector<LoadRequest*> requests,
+    bool prefetch) {
+  if (requests.empty() || (requests.size() == 1 && !prefetch)) {
+    return;
+  }
+  auto load = std::make_shared<SelectiveCoalescedLoad>(
+      input_, ioStats_, groupId_, requests, pool_, options_.loadQuantum());
+  allCoalescedLoads_.push_back(load);
```

Comment:
So the size of allCoalescedLoads_ is always the same as coalescedLoads_? Thanks!